File: dfs_bfs_backtracking/cpp/b2178.cpp
#include <iostream>
#include <string>
#include <vector>
#include <queue>

int N, M;
int fields[100][100];
std::vector<std::pair<int, int>> directions { {1, 0}, {-1, 0}, {0, 1}, {0, -1} };

void bfs() {
	std::queue<std::pair<int, int>> check;
	check.push(std::make_pair(0, 0));
	
	while (check.size()) {
		int row = check.front().first;
		int col = check.front().second;
		if (row == N - 1 && col == M - 1) {
			std::cout << fields[row][col] << std::endl;
			return;
		}
		check.pop();
		for (std::pair<int, int>& direction : directions) {
			int new_row = row + direction.first;
			int new_col = col + direction.second;
			if (new_row == 0 && new_col == 0) { continue; }
			if (new_row >= 0 && new_row < N && new_col >= 0 && new_col < M && fields[new_row][new_col] == 1) {
				check.push(std::make_pair(new_row, new_col));
				fields[new_row][new_col] = fields[row][col] + 1;
			}
		}
	}
}

int main() {
	std::ios_base::sync_with_stdio(false);
	std::cin.tie(NULL);

	std::cin >> N >> M;

	for (int i = 0; i < N; ++i) {
		std::string temp;
		std::cin >> temp;
		for (int j = 0; j < M; ++j) {
			fields[i][j] = temp[j] - '0';
		}
	}
	bfs();
}

File: dfs_bfs_backtracking/python/b2178.py
from sys import stdin

N, M = map(int, input().split())

fields = [list(map(int, list(stdin.readline())[:-1])) for _ in range(N)]

front = -1
rear = 0
queue = [0] * (100 * 100)
queue[rear] = 0

directions = ((1, 0), (-1, 0), (0, 1), (0, -1))

while front != rear:
    front += 1
    row = queue[front] // M
    col = queue[front] % M
    if row == N - 1 and col == M - 1:
        print(fields[row][col])
        break

    for direction in directions:
        new_row = row + direction[0]
        new_col = col + direction[1]
        if 0 <= new_row < N and 0 <= new_col < M and fields[new_row][new_col] == 1:
            if new_row == 0 and new_col == 0:
                continue
            rear += 1
            queue[rear] = new_row * M + new_col
            fields[new_row][new_col] = fields[row][col] + 1
